[mess up] remote_src of copy.py  supports recursive copy of directory

ptux opened the pull request:
##### SUMMARY

fixes #14131 

The followings will be supported.

* push remote src to remote dest with idempotence.
* supports check mode currently.
* supports symlink.

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

module copy.py

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.7.0.dev0 (issue-14131 974076c7cb) last updated 2018/06/07 13:19:30 (GMT +900)
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['~/ansible/library']
  ansible python module location = ~/ansible/lib/ansible
  executable location = ~/ansible/bin/ansible
  python version = 3.6.5 (default, Apr 25 2018, 14:23:58) [GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.1)]

```

##### ADDITIONAL INFORMATION

pilou- committed:
filesystem: add support for f2fs

Closes: #40418

pilou- committed:
filesystem: add f2fs integration tests

Commit:
add feature that remote_src supports recursive copy of directory

Commit:
fixes pep8 issue

jborean93 commented:
Not sure if there is much use in adding recursive copying for remote_src if there is not idempotence or check mode. The user may as well just run a cp command. @abadger I believe you were planning on adding support for this at some point, do you have any feelings about this?

ptux commented:
@jborean93 
thanks for your comment.
I was confused about idempotence.
I did not implement idempotence, because I thought this is copy, not sync.

I would add [check mode] later.

jborean93 commented:
@ptux no worries, `remote_src: no` is idempotent, it checks each file before copying to see if it needs to happen. This is part of the reason why `copy` is slower than just plain SSH commands and not everyone realises this.

ptux commented:
It is annoying.

> it checks each file before copying to see if it needs to happen.

this could happen by using filecmp module of python lib. 
But, it's slow and there is an alternate [synchronize].
 
@abadger
advise, please.

abadger commented:
Copy need to be idempotent and should handle check_mode.  As jborean says,
it is the major reason to use the copy module instead of command: cp

abadger commented:
Also note, filecmp uses file metadata (mode, ownership, etc) as part of
it's comparison.  That may not be appropriate for copy since the user can
set some of those as copy parameters.

bcoca committed:
remove deprecated

pilou- committed:
filesystem: use LANG=C when cmd output is parsed

Commit:
push src to dest with idempotence

Commit:
fix pep8 issues

Commit:
try to fix test failures ["assertion": "copy_result3 is changed"]

Commit:
fix pep8 issue

Commit:
try to support check_mode

Commit:
fix pep8 issues

ptux commented:
@abadger 

Cloud you take a look at the latest commit?
please advise me if there are any problems.

Commit:
fix description

ptux commented:
Is there anybody who will review this PR?

bcoca committed:
fixed test using removed feature

abadger commented on lib/ansible/modules/files/copy.py:
```diff
@@ -78,7 +78,7 @@
       - If C(no), it will search for I(src) at originating/master machine.
       - If C(yes) it will go to the remote/target machine for the I(src). Default is C(no).
-      - Currently I(remote_src) does not support recursive copying.
+      - I(remote_src) support recursive copying.
```
Add version here like:
```
- I(remote_src) supports recursive copying in version 2.7 and higher
```

ptux replied:
got it.

abadger commented:
* Needs integration tests.
* Does this handle permissions?  For instance:
```
- copy:
    remote_src: True
    src: /etc
    dest: ~/backup
    mode: u+rwX,go-rwx
```

ptux commented:
thanks for your comments.

I'll add code to handle permissions.

ptux committed:
try to handler perms

ptux committed:
try to handler perms.

ptux committed:
fix pep8 issues.

ptux commented:
@abadger 

I was coding for handling `mode`,`owner`,`group`.
I noticed it's NOT a good idea that supports `mode` for **remote recursive copy**.
Which I mean who wants to grant SAME permission to dir and file.

So, I  suggest that make **remote recursive copy** support `owner` and `group` only.

Users cloud use `file` module to change permission for dir and file **separately**

ptux committed:
try to handle owner and group, not mode.

ptux committed:
fix pep8 issues

ptux committed:
chown without func

ptux committed:
try to handle chown and symlink

ptux committed:
fix pep8 issue.

ptux commented:
@abadger 

* push remote src to remote dest with idempotence.
* supports check mode currently.
* supports symlink.

review one more time, please.
tell me if there is something wrong or missing.

ptux commented:
ready_for_review

andreaso committed:
Prefer https:// links in the docs site

All the changed urls are availible by way of https://. Most of them
already redirect.

ptux commented:
@abadger 

I guess you have a lot going on.
just a friendly reminder.

bcoca committed:
nicer error message on multiprocessing fail

fixes #43090

Commit:
Properly detect credentials for tower_project

It seemed like it was mostly the wrong variables been looked at, making
it so a git repository could not be created without a credential.

robert-bo-davis committed:
Fix referenced before assignment in role requirements

robert-bo-davis committed:
Fix referenced before assignment in role requirements

abadger commented:
@ptux thanks.  This still neds integration tests.  Probably somewhere around here: https://github.com/ansible/ansible/blob/devel/test/integration/targets/copy/tasks/tests.yml#L1160

ptux commented:
@abadger 
I didn't know how and where to write the test code.
I'll try my best to do it.

abadger committed:
Normalize config from environment as text strings

On Python3, these would be text strings already.  On Python2, we need to
convert them from bytes.

Use a new helper function py3compat to do this.

Fixes #43207

ptux committed:
first try at test

joren485 committed:
Add specific exceptions to except

abadger committed:
Fix config manager to show errors

If there were fatal bugs in this portion of config, they would never be displayed
because config would fail to load and then every^U (Exageration... only
half of every) other part of the code which depended on config to be
loaded would fail before we ever got around to a section of code that
would process UNABLE.

Remove the try except from here so that we are able to debug this code

ptux committed:
make chown support check mode

ptux committed:
make chown support check mode(fix logical bug)

ptux committed:
fix pep issuse.

justjais committed:
Allow DNS bypass for add/remove of host records with nios_host_record (#42934)

* fixes issue 42420

* fixes issue 42420

* fix shippable docs error

* shippable fix and test case add

* shippable fix and test case add

* shippable fix and test case add

* shippable fix

* removing extra assert

* shippable fix

ptux committed:
make sure remote_dir_src not change

ptux committed:
make sure remote_dir_src not change

ptux committed:
check stat of src may be changed by miss

ptux committed:
check stat of src may be changed by miss

felixfontein committed:
Forgot to include validations.

avaika committed:
Leading dot in efs_facts.mountpoint (#35909)

* remove leading dot in efs_facts mountpoint

* remove leading dot in efs_facts mountpoint

* remove leading dot in efs.py

* introduce filesystem_address var in efs* modules and revert changes to mount_point

Qalthos committed:
\r does not round-trip the local socket, escape and restore on the other side (#43507)

joren485 committed:
Use socket.error instead of non-existent socket.Error (#43499)

trishnaguha committed:
Fail module if configure replace fails nxos (#43546)

Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

hemskgren committed:
asa_config doc attach acl to interface (#41253)

+label: docsite_pr
Issue ansible/community#311

ptux committed:
fix issue [ get_url does not change mode when checksum matches ] (#43342)

* fix  #29614

* add change log for #43342

* Cleanup tests and add tests for this scenario


Co-authored-by: ptux

vista- committed:
NCLU - Add example: Changing the hostname and committing on Cumulus Linux (#42872)

* Add example to change the hostname and commit on Cumulus Linux

* Modify phrasing a bit

JacksonIsaac committed:
junos_vlan: Add Vlan ID Range in description (#41855)

wenottingham committed:
Add some requirements to the plugin docstrings.

ptux committed:
testcase [make sure remote_dir_src not change] (fix typo)

ptux committed:
testcase [make sure remote_dir_src not change] (fix typo)

ptux committed:
testcase [make sure remote_dir_src not change] (fix typo)

ptux committed:
testcase [make sure remote_dir_src not change] (fix typo)

ptux committed:
testcase [make sure remote_dir_src not change] (add test items)

ptux committed:
testcase [make sure remote_dir_src not change] (add test items)

ptux committed:
testcase [make sure remote_dir_src not change] (delete test item md5)

ptux committed:
testcase [make sure remote_dir_src not change] (delete test item md5)

ptux committed:
testcase [make sure remote_dir_src not change] (fix typo)

ptux committed:
testcase [make sure remote_dir_src not change] (fix typo)

ganeshrn committed:
Fix typo in cliconf get_config (#43553)

ptux committed:
prepare for test (just modify style )

ptux committed:
prepare for test (just modify style )

ptux committed:
testcase1 [remote_dir_dest not exists] (add test items)

ptux committed:
testcase1 [remote_dir_dest not exists] (add test items)

ptux commented:
@abadger 

I'll continue writing the test code.
cloud you take a look at it at this monment?

advise me if I'm going in the wrong  **DIRECTION**.

ptux committed:
testcase1 [remote_dir_dest not exists] (delete non-logical test item )

ptux committed:
testcase1 [remote_dir_dest not exists] (delete non-logical test item )

ganeshrn committed:
Fix netconf plugin set_option (#43589)

* To enable set config option with `Netconf` plugin
  inherit `NetconfBase` class from `AnsiblePlugin` parent class

Commit:
improved vmss documentation (#43590)

ganeshrn committed:
Raise exception if command timeout is triggered (#43078)

* Raise exception if command timeout is triggered

Fixes #43076

If persistent connection timeout is triggered, riase
exception which will be send over socket to module code
instead of silently shutting down the socket.

* Fix CI failure

* Fix review comment

* Fix CI failure

* Fix review comment

* Fix review comment

Commit:
Replaced old sample in lb (#43591)

mjmayer committed:
Fix comparison of priority (#43329)

The existing rule priority comes from aws as a string. It is then
compared to the new rule priority, which is defined as an int. This change
casts the new rule priority as a string making the comparison work. The
reason to cast it as a string rather than an int is used because a priority
can also be set to 'default'. When trying to case 'default' as an int, it creates
an error.

wookietreiber committed:
improves pam_limits documentation (#43526)

- the man page is actually limits.conf
- uses code highlighting for these

deiga committed:
Adds example of using wildcard domain in pam_limits (#43391)

* Adds example of using wildcard domain

* Fixes W293

* Slight formatting changes on the examples

ruzickap committed:
Removing required from documentaion of name parameter (#43530)

* Removing required from documentaion of name parameter

* extra text to make clearer when name is required

seuf committed:
Moved grafana 5 dashboard compatible changelog to fragments dir (#43593)

trishnaguha committed:
cliconf refactor eos, nxos fix (#43596)

* cliconf refactor eos and nxos fix

Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

* move nxos replace check to plugins

Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

* address review comment

* fix shippable

rrey committed:
Terraform: add workspace support for switching and removing workspaces (#43489)

From terraform documentation:
```
The persistent data stored in the backend belongs to a workspace. Initially the backend has only one workspace, called "default", and thus there is only one Terraform state associated with that configuration.

Certain backends support multiple named workspaces, allowing multiple states to be associated with a single configuration. The configuration still has only one backend, but multiple distinct instances of that configuration to be deployed without configuring a new backend or changing authentication credentials.
```

This patch introduces the `workspace` parameter in the terraform module. The module will select
the workspace is it does not exists, or simply select it if it exists.

Fixes #43134

Add 'purge_workspace' parameter and handle the workspace context

The `purge_workspace` parameter allows to remove a workspace when asking for state = absent.
It allows to leave a clean state file without empty workspaces if the parameter is true.

Also adding the support of a workspace context that allows to restore the workspace
when that was active when the module started.

pilou- committed:
Tower modules: move HAS_TOWER_CLI check in module_utils and minor improvements (#39809)

* tower_* modules: move HAS_TOWER_CLI in TowerModule

Besides this change allows to define other common parameters such as
mutually_exclusive.

* tower_*: config file can not be used with auth params

* tower module_utils: remove useless call to expanduser

'path' type: expanduser & expandvars are automatically called

sebastianw committed:
Fix ini variables for netconf and network_cli connections (#43599)

bcoca committed:
more useful messages when module failure (#43576)

* more useful messages when module failure

specially if the 'interpreter' is not found

* 1 less var

* shebang can be none

* remove typoes

Commit:
Clarify usage of encrypt arg (#43330)

According to the do_encrypt interface, encrypt arg should be the hash method name used for encrypting returning password. But in the doc and lookup code it's a boolean flag, correct it to string.

vfoucault committed:
Support for postgresql default privileges (#32987)

* Support for postgresql default privileges

fix the following issues:

* #29701
* #23657

* The ALTER DEFAULT PRIVILEGES is implemented with type 'default_privs'
* Added a Query Builder for simplification
* Some minor lint

* Fixed Lint Issue in doc

Fixed misspelled method name

* Removed the damned empty space on line 243 ! (within the doc) x|

* Kept Compat in string interpolation for old beloved python 2.6

bcoca committed:
added version info and removed 'bare' exception

Everspace committed:
Implement signed_request for sigV4 requests (#42758)

* Implement signed_request for sigV4 requests

* Correct linting errors

* More linting changes. Correct import.

* Final linting fix for inline comments

* Correct import of urllib.parse

* Update copyright and shebang line

* Remove shebang

* Put boto3 requirement. Abtract out get_aws_key_pair for module consumption.

* Dummy out unused region variable.

* Handle Boto3 ImportError

* - implement get_aws_credentials_object with willthames suggestion
- Handle session_token

* Make quote style consistant

* Chop arugment line up

* Correct indent

hi117 committed:
Add scope parameter to systemd (#40179)

* Added changes for Issue #38828, adds scope paramater to systemd module

* Removed description for old paramater

* Added version_added field for new option

* Readded the user paramater as a deprecated paramater

* Changed version for the scope paramater since I missed the release window

ptux committed:
fix non-logical for link follow

ptux committed:
fix non-logical for link follow

ptux committed:
fix non-logical for follow

ptux committed:
fix non-logical for follow

ptux committed:
testcase1 [remote_dir_dest not exists] (fix test item)

ptux committed:
testcase1 [remote_dir_dest not exists] (fix test item)

ptux committed:
testcase1 [remote_dir_dest not exists] (fix test item)

ptux committed:
testcase1 [remote_dir_dest not exists] (fix test item)

ptux committed:
testcase1 [remote_dir_dest not exists] (fxi test item)

ptux committed:
testcase1 [remote_dir_dest not exists] (fxi test item)

ptux committed:
testcase1 [remote_dir_dest not exists] (fix typo for test item)

ptux committed:
testcase1 [remote_dir_dest not exists] (fix typo for test item)

Commit:
add local directory option to nxos_file_copy (#43277)

* add local directory option to file_copy pull

*  shippable fix

dodwmd committed:
Allow for the specification of a rhsm_repo_ca_cert if changing baseurl (#43343)

* Allow for the specification of a rhsm_repo_ca_cert if changing baseurl

If changing the baseurl we should allow for a ca cert to be updated from redhat-uep.pem

* Fixing documentation section

* added version to option rhsm_repo_ca_cert

* got rid of extra CR

rrey committed:
modules/terraform: use -no-color option in workspace commands (#43628)

Commit:
Refactoring to persistence connection BGP, factory, reload, save, showrun modules (#43534)

* Refactoring to persistence connection BGP, factory, reload, save, showrun modules

* Refactoring methods from Util to module file

* Removing BGP Utility methods

* Adding to errors that need to be ignored

NilashishC committed:
Fix argument error in DCI (#43634)

Nifi committed:
Fix telnet action plugin type error (#43609)

* Fix telnet action plugin type error

* removed blank line too much (pep8 E303)

rrey committed:
modules/terraform: fix -no-color position in workspace command (#43644)

abadger committed:
Only print warning when ansible.cfg is actually skipped (#43583)

Only print warning when ansible.cfg is actually skipped

* Also add unittests for the find_ini_config_file function
* Add documentation on world writable current working directory
  config files can no longer be loaded from a world writable current
  working directory but the end user is allowed to specify that
  explicitly.  Give appropriate warnings and information on how.

Fixes #42388

sivel committed:
Fix invocation output. Fixes #43447 (#43542)

sdodsley committed:
Update Pure Storage FA pgsnap module to allow volume restore from snap (#42774)

sdodsley committed:
Add support for host personality in purefa_host module. (#42976)

This is recommended when ActiveCluster is supported in the FlashArray

jborean93 committed:
win_chocolatey: fix regression around using all as a package name (#43483)

agrimal committed:
Fix bug #32437 (#38166)

abadger committed:
changelog for lxd idempotence fix

jborean93 committed:
win-async: fix race condition in async run (#43691)

kbreit committed:
Meraki/doc response vlan (#43669)

* Add proper response documentation to meraki_vlan

* Add fixed IP assignemnts and reserved IP range response docs

* Fix syntax error

* Remove duplicate reservedIpRanges documentation

jborean93 committed:
Increase win_async_wrapper fire and forget time check for more stabilisation

ptux committed:
fix follow logic

ptux committed:
fix follow logic

ganeshrn committed:
Update junos cliconf plugin (#43643)

* Update junos cliconf plugin

Fixes #39056

Refactor junos cliconf plugin api and other
minor changes

* Fix CI issue

* Fix CI failure

LindsayHill committed:
Typo in edit_banner (#43668)

Looks like `send_commad` was mistakenly added in https://github.com/ansible/ansible/pull/43203. Should be `send_command`

rmudingay committed:
junos_user: additonal sshkey types added (#43685)

ptux committed:
testcase1 [remote_dir_dest not exists] (fix test item)

ptux committed:
testcase1 [remote_dir_dest not exists] (fix test item)

Akasurde committed:
delegate_to: accept only string for delegation host (#43473)

delegate_to parameter in task only accepts string,
this fix will error out if other datatypes are provided instead of
string.

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

ptux committed:
testcase1 [remote_dir_dest not exists] (fix test item)

ptux committed:
testcase1 [remote_dir_dest not exists] (fix test item)

individuwill committed:
change raising errors example to raise error (#43664)

Example instantiates an AnsibleError which derives from Exception but doesn't actually raise it like intended. This is misleading as it's not clear without examining the code for AnsibleError to know that it's not some function which would raise the exception automatically.

gustavomcarmo committed:
ldap_entry bugfix (#43620)

tiraboschi committed:
ovirt: ovirt_hosts_facts: add all_content (#43525)

Add all_content parameter also to ovirt_hosts_facts
to fetch the complete list of of details.

Qalthos committed:
cli_command module (#42916)

* Create cli_command module to do direct cliconf calls

* Update edgeos cliconf signature to match.

* cli_command is cli-only

* Add tests

azaghal committed:
Fix introduction version for parameter "local" in module "group" (#43719)

Parameter "local" has been introduced to the "group" module in version 2.6, not 2.5.

agaffney committed:
Force all args to strings for 'command' (fixes #43732)

dagwieers committed:
ACI: Add ability to enable debug for tests (#43744)

DavidWittman committed:
Add cap_drop to docker_container module (#36889)

Closes #29578

felixfontein committed:
ACME: support for TLS-ALPN-01 (#42158)

* Added support for TLS-ALPN-01 verification.

* Unrelated commit to re-trigger tests.

* Added test for TLS-ALPN-01.

* Try to remove to_bytes in the hope that binary data survives in Python 2.

* Using Base64 encoding for TLS-ALPN-01 value.

willthames committed:
Allow tower_inventory_sources params to be False (#43749)

Ensure that params set to False are respected, rather than ignored.

Akasurde committed:
VMware: added missing facts vmware_drs_rule_facts (#42992)

vmware_drs_rule_facts was missing details about vm_group and host_group
and their respective memeber names. This fix adds those details and updates documentation.

Fixes: #42980

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

Akasurde committed:
VMware: new module : vmware_guest_boot_facts (#43240)

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

Akasurde committed:
VMware: confine search for network till datacenter level (#40382)

If there are multiple distributed virtual portgroup network with same name
in different datacenter, vmware_guest module used to choose first DVPG irrespective
of the datacenter. This caused problem if two datacenter has same name for DVPG,
vmware_guest module failed with error "The object or item referred to could not be found."

This fix adds check to search for network (Distributed Virtual Portgroup)
till datacenter level. This avoids selection of same name DVPG from other datacenter.

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

AlexisLessard committed:
Clarifications of parameters in yum_repo module (#38453)

* Clarifications of parameters in yum_repo module

Added a note defining where the "name" parameter of the module will appear in the repo file and note explaining that the description parameter of the module is actually the name parameter in the repo file. It might help people transform their existing yum repository files in ansible managed repos.

IlnarSelimcan committed:
Update intro_installation.rst (#39381)

Apparently the subsection on running from source has been moved around and now it appears *below*.

+label: docsite_pr

toadjaune committed:
Correct a typo in pip module error message (#43784)

Spredzy committed:
Vultr: Ensure facts works when no resource exists (#43603)

Vultr API is being inconsisten in what it returns. An empty list when no
resources exists, but a dict of dict when they do. The case needs to be
handled so the module do not fail. An extra test has been added.

Akasurde committed:
Misc typo fixes in function docs (#43779)

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

jerryrelmore committed:
Corrected spelling of 'formatted' (#43785)

curry9999 committed:
[docs] required arguments: name (#43758)

<!--- Your description here -->

+label: docsite_pr

willricardo committed:
[aws] Allow CloudWatch Log Group retention to be changed when log group exists (#43752)

* fix bugs #42028 and #43735

curry9999 committed:
To improve readability, we added a line feed. (#43764)

* A line break was added.

<!--- Your description here -->

+label: docsite_pr

* "blank line contains whitespace" was fixed.

curry9999 committed:
I changed the document of win_hostname.py. (#43761)

* Host name is contrary to naming convention

<!--- Your description here -->

+label: docsite_pr

* removed yaml separators

dagwieers committed:
ACI: Make querying links and nodes possible (#43441)

This functionality was not considered when the module was written, but
there's no reason why it shouldn't be supported.

We had to rework the query string construction and object filtering.
This new functionality allows to filter on arbitrary keys and supports
None values.

This PR fixes various issues with the existing framework, including
querying specific objects using construct_url_4 (i.e.
aci_epg_to_contract and aci_static_binding_to_epg)

yungezz committed:
add azure_rm_webapp (#40005)

* add azure_rm_webapp

* fix lint error

* resolve comments

* fix lint

* fix lint

* fix lint

* fix lint

* fix test failure

* fix test

* fix test

* fix test

* refine ux

* resolve comments

* fix lint and test

* fix lint

* code cleanup

* combine windows_frameworks/linux_framework to one

* fix lint in doc

* fix l int

* fix lint

* fix bug

* fix bug

* fix bug

* fix bug

* fix syntax error

* fix lint error

* fix bug of wrong indent

* fix test issue

* update java framework parameter

* fix lint error

* resolve comments

* fix lint

* resolve comments

* fix bug

* update test

* fix lint

* fix test

yungezz committed:
add azure_rm_appserviceplan module (#40906)

* add azure_rm_appserviceplan module

* fix lint in test

* fix lint

* fix test

* fix test syntax

* resolve comments

* fix test

yuwzho committed:
Fix nsg cannot add rule with  purge_rules false (#43699)

pilou- committed:
improve 'service' integration tests (#43655)

* service tests: check that daemon is really running

(spoiler: it isn't)

* service tests: add PIDFile directive in systemd unit

* service tests: check 'changed' too

* service tests: fix indentation & use changed test

* service tests: #16694 has been fixed a long time ago

* service tests: refactor

- always execute cleaning tasks
- move tests tasks in a dedicated file

* service tests: add test for #42786

* service tests: display value of ansible_service_mgr

* service tests: allow to run tests twice in a row

stop & disable ansible test service

* service tests: 'pattern' value must be a substring

'pattern' parameter is poorly named

* service tests: check ansible_test service status

* service tests: test daemon must handle SIGHUP

because 'initctl reload' sends SIGHUP, otherwise test daemon stops when
receiving the signal

* service tests: remove upstart override file too

and check that files were removed using raw module and stat command

trishnaguha committed:
revert #43546 nxos terminal plugin (#43806)

Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

yuwzho committed:
refine aks  document (#43810)

ganeshrn committed:
Update eos clcionf plugin (#43809)

* If the remote host supports session
  capability set the `supports_generate_diff`
  flag to false which identifies if config diff should
  be generated within module or not

Akasurde committed:
VMware: Update example (#43786)

* Unified examples in all modules
* validate_certs is now 'no' instead of 'False'
* delegate_to changes

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

acalm committed:
make lockfile destination settable and update doc (#42795)

* make sure lock file is available for others when unlocking

* add tmpdir option and updated documentation

LindsayHill committed:
New networking module: voss_command (#43741)

* new voss_command module

* Removed incorrect version_added line

sward1557 committed:
Added junos_config note about override set format (#43608)

hlrichardson committed:
network/exos: add exos_facts module (#43210)

Add exos_facts module. Known limitations at this time include:
   - Interface MTU is not reported.
   - Only primary interface IP is reported.

Add basic unit tests for the exos_facts module.

An EXOS CLI prompt can be prefixed with '! ' (shutting down), '* '
(running configuration does not match saved configuration), and
can include various status tokens within parentheses after these
prefixes. Update prompt regex to accept valid CLI prompts.

lingxiankong committed:
Add os_loadbalancer module for OpenStack Octavia service (#42552)

lingxiankong committed:
Add os_listener module for OpenStack Octavia (#42604)

mattclay committed:
Correct task args in azure_rm_managed_disk test.

mattclay committed:
Fix PEP 8 issue in exos_facts.

openstacker committed:
Add OpenStack Magnum cluster template module (#42654)

Commit:
Lenovo size reduce cnos.py of util (#43823)

* Changing Lenovo Inc to Lenovo and update License file to be consistent.

* Moving code in util file to module file. This is done as per a review comment

* Update cnos_vlag.py

* Update cnos_vlan.py

* Update enos_config.py

Qalthos committed:
Network *_config action plugin updates (#43838)

* Compatibility fixes for ${network_os}_config action plugins

sivel committed:
Ensure removed_in is StrictVersion before comparing (#43835)

* Ensure removed_in is StrictVersion before comparing

* Catch ValueError with StrictVersion on incompatible versions

dagwieers committed:
Fix all ACI examples to use delegate_to (#43439)

This is to avoid users connecting to the APIC using SSH, which is likely
to fail. In the ACI documentation (linked from every module) we go into
more details on how people are supposed to work with the modules..

ericwb committed:
Create a template for security related bugs (#43836)

* Create a template for security related bugs

This template serves as a redirect for people who might not know where to open security related issues.  It guides them to email security@ansible.com instead of opening an issue here.

* Update security_bug_report.md

andreaso committed:
Fix link description typo in tools documentation (#43828)

curry9999 committed:
To improve readability, we added a line feed. (#43801)

* 	modified:   google/gcp_compute_backend_bucket.py
	modified:   google/gcp_compute_backend_service.py
	modified:   google/gcp_compute_forwarding_rule.py
	modified:   google/gcp_compute_global_forwarding_rule.py
	modified:   google/gcp_compute_image.py
	modified:   google/gcp_compute_instance.py
	modified:   google/gcp_compute_instance_group.py
	modified:   google/gcp_compute_instance_group_manager.py
	modified:   google/gcp_compute_instance_template.py
	modified:   google/gcp_compute_route.py
	modified:   google/gcp_compute_subnetwork.py
	modified:   google/gcp_compute_target_http_proxy.py
	modified:   google/gcp_compute_target_https_proxy.py
	modified:   google/gcp_compute_target_ssl_proxy.py
	modified:   google/gcp_compute_target_tcp_proxy.py
	modified:   google/gcp_compute_url_map.py
	modified:   google/gcp_container_node_pool.py
	modified:   google/gcp_dns_resource_record_set.py
	modified:   google/gcp_pubsub_subscription.py
	modified:   google/gcp_storage_bucket_access_control.py

* 	modified:   lib/ansible/modules/cloud/amazon/aws_ses_identity.py
	modified:   lib/ansible/modules/cloud/amazon/route53_facts.py
	modified:   lib/ansible/modules/cloud/cloudscale/cloudscale_server.py
	modified:   lib/ansible/modules/network/aos/_aos_logical_device.py
	modified:   lib/ansible/modules/network/aos/_aos_rack_type.py
	modified:   lib/ansible/modules/network/aos/_aos_template.py
	modified:   lib/ansible/modules/network/cumulus/nclu.py
	modified:   lib/ansible/modules/web_infrastructure/ansible_tower/tower_job_launch.py
	modified:   lib/ansible/modules/web_infrastructure/ansible_tower/tower_job_wait.py

akira6592 committed:
Corrects module name in net_get module doc (#43816)

gundalow committed:
remove gundalow from Network issues (#43849)

dagwieers committed:
Support dash-separated leaf IDs (as documented) (#43846)

dagwieers committed:
aci_fabric_node: Fix idempotency (with newer ACI ?) (#43848)

gundalow committed:
Don't notify privateip on network changes (#43859)

Akasurde committed:
VMware: remove unused imports (#43804)

Idea taken from https://github.com/ansible/ansible/pull/43402

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

antoineco committed:
azure_rm_deployment: collect tags from existing Resource Group (#26104)

alexjbush committed:
azure_rm_loadbalancer_facts.py: list() takes at least 2 arguments fix (#29046) (#29050)

jborean93 committed:
ansible-test: Create public key creating Windows targets (#43760)

* ansible-test: Create public key creating Windows targets

* Changed to always set SSH Key for Windows hosts

EtienneDeneuve committed:
add group_by_os_family in azure dynamic inventory (#40702)

trishnaguha committed:
set supports_generate_diff to False vyos (#43873)

Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

Akasurde committed:
Fix unreferenced msg from vmware_host (#43872)

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

NilashishC committed:
Add source interface and use-vrf features (#43418)

pilou- committed:
[doc] import_role: mention version from which behavior changed and fix some typos (#43843)

* [doc] fix some typos

* [doc] import_role: mention version from which behavior changed

Qalthos committed:
Remove unnecessary features from cli_command (#43829)

* Remove wait_for & associated params

* Upgrade command dict to top-level, remove more wait_for accessories

* We don't need all this anymore

* Update docs

* Update tests to new argspec

* Be a little more explicit about sendonly

Also remove reference to the word provider.

* Add example of prompt and answer

mbusbee committed:
VMware: datastore selection (#35812)

This fix adds additional parameter 'datastore'. This parameter
provides a flexibility to specify custom datastore or datastore cluster
to deploy new virtual machine from template (which is located in different
datastore or datastore cluster from virtual machine's datastore or datastore
cluster).

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

pilou- committed:
Docker integration tests: factorize setup (#42306)

* Factorize docker_secret & docker_swarm tests setup

* failure are only expected inside docker container

* docker_swarm: enable RHEL builds

remixtj committed:
New module for managing EMC VNX Block storage (#42945)

* New module for managing EMC VNX Block storage

With the module emc_vnx_sg_member users can add or remove luns from
existing storage groups.

This module has been developed in couple with emc_vnx_mv_promote for
disaster recovery process automation, but can be used by itself.

mattclay committed:
Fix download URL for yum integration test.

LindsayHill committed:
added platform guides for NOS and VOSS (#43854)

ryansb committed:
Fix spelling of ansbile to ansible (#43898)

dagwieers committed:
Fix typo

mattclay commented:
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/77323/46/tests

gdpak committed:
openvswitch_db : Handle column value conversion and idempotency in no_key case (#43869)

* change column value to string. handle idempotency in no key case

* fix unit test failures in baseline

jborean93 committed:
win_domain modules: ensure Netlogon service is still running after promotion (#43703)

ganeshrn committed:
Update iosxr cliconf plugin (#43837)

* Update iosxr cliconf plugin

Fixes #39056

*  Update iosxr cliconf plugin
*  Modify iosxr module_utils code to support
   refactored cliconf plugin api's
*  Other minor changes

* Fix unit test failure

* Update ios, eos, nxos plugin for diff

* Fix review comment

ganeshrn committed:
Fix eos_command integration test failures (#43922)

Qalthos committed:
Prevent data being truncated over persistent connection socket (#43885)

* Change how data is sent to the persistent connection socket.

We can't rely on readline(), so send the size of the data first. We can
then read that many bytes from the stream on the recieving end.

* Set pty to noncanonical mode before sending

* Now that we send data length, we don't need a sentinel anymore

* Copy socket changes to persistent, too

* Use os.write instead of fdopen()ing and using that.

* Follow pickle with sha1sum of pickle

* Swap order of vars and init being passed to ansible-connection

trishnaguha committed:
nxos_vlan refactor to support non structured output (#43805)

* nxos_vlan refactor to support non structured output

Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

* unittest fix

Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

* minor fixes

Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

* use check_rc

Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

* address review comment

Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

* remove additional return statement

* address Nate's review

Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

sivel committed:
Remove use of simplejson throughout code base (#43548)

* Remove use of simplejson throughout code base. Fixes #42761

* Address failing tests

* Remove simplejson from contrib and other outlying files

* Add changelog fragment for simplejson removal

ssemyonov committed:
Introduce `force` and `verbose` options in serverless module (#43947)

* Introduce `force` and `verbose` options in serverless module

* Add "version_added: 2.7" for the new `force` and `verbose` options

carchi8py committed:
Adding Common files for NetApp ElementSW release (#43727)

* MVP2 Post ElementSW OSRB sync

* Revert "MVP2 Post ElementSW OSRB sync"

This reverts commit c13db2ad962cd56bffce052c2891c558a2240c72.

abadger committed:
Remove call to parse_returned_data when removing tmpdir

_parsed_return_data should only be used with the return from a module.
This location was invoking a remote shell command rather than a module
so we don't want it here.

agaffney committed:
Display correct task title with strategy 'free' (fixes #43950)

bcoca committed:
handle env exception in gathering even no inject (#43569)

* handle env exception in gathering even no inject

(cherry picked from commit 60e3b9b3527daea6d9fb75a80cc14b35923748f3)

Commit:
pkg_mgr: fixed apt_rpm detection (#43769)

Instead of checking the distribution name (which apparently is tricky to find out)
check if /usr/bin/apt-get is managed by RPM.

Fixes #43539

bcoca committed:
fix alt linux detection for new versions (#43723)

they dropped the Linux so now it only shows as ALT, it should still be backwards compatible
pkg_mgr detection relies on `Altlinux` string, so properly setting os_distribution should take care of it as side effect
fixes #43539

gundalow committed:
iosxr_system always syntax error (#43977)

jairojunior committed:
Add botmeta for team jboss. (#43961)

LindsayHill committed:
new nos_command module (#43056)

spmp committed:
Add enable_ip_forwarding option to azure_rm_networkinterface; Fixes #43276 (#43335)

* Merge again trickily similar Accelerated networking and IP forwarding

* Add type to enable_ip_forwarding documentation

* Fix merge error

* auth to auto

* azure_rm_networkinterface: remove auth_source from tests

* azure_rm_networkinterface: remove spurious auth_source from test

* azure_rm_networkinterface: Revert formatting on test

* azure_rm_networkinterface: Correct indentation

Qalthos committed:
replace errant use of 'response' with 'answer' (#43985)

mattclay committed:
Fix path handling in hacking/env-setup.

resmo committed:
cs_pod: workaround for 4.11 API break (#43944)

remixtj committed:
docs: Fixed wrong storage platform on module docs for EMC VNX (#43899)

Spredzy committed:
module_utils/vultr.py: Ensure comparison is accurate (#43298)

In query_resource_by_key(), there is an equal comparison that is made to
know if the object we are looking for is present. Due to type difference
this comparison doesn't always retrieve true, even when it should.

This is due to the fact that the value in r_data dict are of type
unicode, while the other can be of type int, float,... .

```
>>> a = u'1'
>>> type(a)
<type 'unicode'>
>>> b = 1
>>> type(b)
<type 'int'>
>>> a == b
False
>>> str(a) == str(b)
True
```

Hence the values, for comparison purposes, are casted into strings.

ptux committed:
fix non-logical misstake

ptux committed:
fix non-logical misstake

ptux committed:
fix TypeError from ubuntu

ptux committed:
fix TypeError from ubuntu

ptux committed:
add testcase2 remote_dir_dest exists, left only subdir and file exists

ptux committed:
add testcase2 remote_dir_dest exists, left only subdir and file exists

ptux committed:
fix tpye error for py3

ptux committed:
fix tpye error for py3

ptux committed:
add testcase3 remote_dir_dest exists, diff files in common dirs

ptux committed:
add testcase3 remote_dir_dest exists, diff files in common dirs

ptux committed:
single quote missed

ptux committed:
single quote missed

ptux committed:
fix misspelled module name

ptux committed:
fix misspelled module name

ptux committed:
fix wrong path

ptux committed:
fix wrong path

ptux committed:
fix misspelled var

ptux committed:
fix misspelled var

ptux committed:
fix testcase3

ptux committed:
fix testcase3

ptux committed:
fix type error bytes str

ptux committed:
fix type error bytes str

ptux committed:
fix type error bytes str

ptux committed:
fix type error bytes str

ptux committed:
add testcase4 Copy the directory on remote without chown

ptux committed:
add testcase4 Copy the directory on remote without chown

ptux committed:
improve testcase4

ptux committed:
improve testcase4

ptux committed:
fix syntax error

ptux committed:
fix syntax error

ptux committed:
fix type error

ptux committed:
fix type error

ptux commented:
@abadger 
@IPvSean 
@mattclay 

ready to review!
anybody, please.

ptux committed:
Merge branch 'issue-14131' of https://github.com/ptux/ansible into issue-14131
